### ISSUE TITLE
Designate some frames as "null" frames with no animals

### ISSUE THREAD

**Issue by olinesn** [murthylab/sleap#640]
Is there a way to designate some frames as frames that don't have animals, but still include them in the training set? For example, to teach the model "my gloved hand is not a mouse." Often I'll have a few new tracks generated just from reaching my hand into the cage to place down a mouse, which leads to a lot of necessary cleanup.
![Null_frames](https://user-images.githubusercontent.com/69451171/152675957-058ad8bc-158a-4a88-8a72-eb4fe499fa30.png)

As far as I know, if I delete the skeleton below, then the frame is not included in the training set.

Thanks!

**Comment by talmo** [comment]
Definitely added the to-do list. Just out of curiosity: have you tried adding an instance to one of these frames and just marking all of the nodes as not visible?

**Comment by olinesn** [comment]
Thanks - didn't know that was an option. If that's the right way to do it, I'll start doing that!

**Comment by roomrys** [comment]
Since we seem to have gotten a lot of cases where people are marking all nodes as not visible, wanted to state here (at the root) that SLEAP will filter out all empty instances (with all non visible points). Labeling an instance as entirely occluded will not do anything (other than use-up your time):
https://github.com/talmolab/sleap/blob/833c2d5bdcf4cfef4e0adc7569b8e2245494a8fa/sleap/nn/data/providers.py#L37-L59

**Comment by olinesn** [comment]
> Since we seem to have gotten a lot of cases where people are marking all nodes as not visible, wanted to state here (at the root) that SLEAP will filter out all empty instances (with all non visible points). Labeling an instance as entirely occluded will not do anything (other than use-up your time):
> 
> https://github.com/talmolab/sleap/blob/833c2d5bdcf4cfef4e0adc7569b8e2245494a8fa/sleap/nn/data/providers.py#L37-L59

Got it, thanks for the update!

**Comment by calebvogt** [comment]
Has there been any movement on this front? Am I correct in understanding that in a frame which contains no animal, putting in an instance and then labeling all nodes as "not visible" will not incorporate that frame in the training set as "background" material, and is therefore a waste of time? @talmo's comment above (and I think in other comments I have seen) seems to indicate that this "invisible" instance method on an empty frame is viable?

**Comment by olinesn** [comment]
As far as I know, this is still the case. We actually had a problem in the lab this week because of this, SLEAP is generating tons of low-scoring skeletons when we reach into a cage to place a mouse. Our solution in the near term is to take some video where the animal is in frame, but we also wave our hand around, and then label the animal in those frames so that the model learns to not detect the gloved hand. A way to 'designate' null frames would be nice though, and a way to quickly update frames from a new context.

**Comment by olinesn** [comment]
@talmo any ideas? <3

**Comment by talmo** [comment]
This is definitely one that we should prioritize, I agree.

Technically, in terms of the backend, this is pretty doable by just getting rid of [the filtering in `LabelsReader`](https://github.com/talmolab/sleap/blob/1151a95cd699904d5ba28b591bcb21c2faf2ca89/sleap/nn/data/providers.py#L55-L56) and using the existing [`Labels.negative_anchors`](https://github.com/talmolab/sleap/blob/1151a95cd699904d5ba28b591bcb21c2faf2ca89/sleap/io/dataset.py#L422) attribute which is designed to store this. The original vision was to able to mark ROIs in the background, but we never finished implementing it.

We had a UX consideration that we wanted to address before doing this: how do you "mark" a frame as negative, and how do you ensure that you don't introduce false negative labels into the dataset?

If a user accidentally marks a frame as a negative example, this is going to introduce a very strong (and incorrect) error signal to the loss that will cause all sorts of weird behavior that will be hard to troubleshoot. For example, how do we find negative frames to check if one was added by accident?

We figured that one way to mitigate this would be to have a GUI element that lists all labeled frames with some filtering/sorting options. #736 is tracking this, but as you can tell, it's been a while and we haven't really converged on a good solution.

Maybe a better short-term workaround would be to have a dedicated panel/table for negative frames? Suggestions on the UX for this feature would be appreciated!

**Comment by olinesn** [comment]
Talmo thanks for the input. I like the idea of a dedicated panel where the user can purposefully designate some 'suggested frames' but for frames that would be included in the training set as null frames. It would be their responsibility to ensure no false negatives, though there could be a way for the software to 'doublecheck' by running inference on those frames and alerting if anything scores over 0.5.

**Comment by skyshi0** [comment]
Hi, I wonder if anyone @talmo @roomrys @gitttt-1234  could comment on the current workaround on this question about how to improve the model’s performance on false positives? I tried adding more “invisible instance”, meaning when the model predicts a wrong object as target instance I set all the predicted nodes to invisible, but it seemed to make the models very confused if those frames are a lot. I had an opened issue here [talmolab/sleap · Discussions · GitHub](https://github.com/talmolab/sleap/discussions/2445) and I appreciate any comments on that

**Comment by davorvr** [comment]
> @skyshi0:
> I tried adding more “invisible instance”, meaning when the model predicts a wrong object as target instance I set all the predicted nodes to invisible

@skyshi0 I tried that too! Currently, I'm using a script where I find these frames (which have exactly one instance with all points invisible), delete the instance, and set the negative flag to True:

```python3
import os
import sleap_io as sio

labels = sio.load_file("labels.v001.pkg.slp")

for idx, fr in enumerate(labels.labeled_frames):
    if len(fr.instances) == 1 and not any(p[1] for p in fr.instances[0].points): # p[1] is the "visible" property of a point
        labels.labeled_frames[idx].instances = list()
        labels.labeled_frames[idx].is_negative = True

sio.save_slp(labels, "labels_neg.v001.pkg.slp")
```

---

> @talmo: 
> We had a UX consideration that we wanted to address before doing this: how do you "mark" a frame as negative

@talmo @roomrys @gitttt-1234 Regarding the UX, it seems both @skyshi0 and I spontaneously tried the same thing to label frames as negative: add an instance, and set all its points as hidden/invisible, so this might be the way to go. I'll try to analyse.

I can identify **three concerns**:

1. Regardless of the UI, how clean-cut is the criterion itself?
2. Does this UI method of labelling frames as negative increase the likelihood of user error?
3. Is this UI method of labelling negative frames intuitive?

##### Concern 1.
If the assumption stands that a negative frame should have zero visible bodyparts and all frames that have zero visible bodyparts are negative frames, the following criterion seems straightforward, accurate, and easy to implement for automatic flagging of negative frames: if there is one or more instances and zero visible points across all of them, then and only then a frame is labelled as negative. This is also an easy test to put in label QC and `sio fix` (negative frame with one or more visible points being flagged as a point of concern).

##### Concern 2.

**_Assumption:_** Adding an instance and then hiding all of its points is a deliberate user action. If so, it follows that:
**_Conclusion:_** Mislabelling a frame as negative in this manner is just as likely as any other labelling errors (e.g. switching symmetrical bodyparts, like left/right ears) - **no increase in the likelihood of erroneous labels**

##### Concern 3.

This is difficult to assess, but the first signal may be that two people independently tried to use the GUI in this way. However, if so, this would be pretty good news for Concern 2 as well. Tying this into Concern 2:

**_Assumption 1:_** If this method won't increase the likelihood of erroneous labels, and
**_Assumption 2:_** If this method is intuitive, then it follows that:
**_Conclusion:_** This way of labelling negative frames is **sufficient** in the GUI, with SLEAP-IO as a remaining, advanced option for more control.

---

Though I have no experience in UX, these questions are always super interesting to think about and I'm very curious to hear what you think.

**Comment by skyshi0** [comment]
> > [@skyshi0](https://github.com/skyshi0):
> > I tried adding more “invisible instance”, meaning when the model predicts a wrong object as target instance I set all the predicted nodes to invisible
> 
> [@skyshi0](https://github.com/skyshi0) I tried that too! Currently, I'm using a script where I find these frames (which have exactly one instance with all points invisible), delete the instance, and set the negative flag to True:
> 
> import os
> import sleap_io as sio
> 
> labels = sio.load_file("labels.v001.pkg.slp")
> 
> for idx, fr in enumerate(labels.labeled_frames):
>     if len(fr.instances) == 1 and not any(p[1] for p in fr.instances[0].points): # p[1] is the "visible" property of a point
>         labels.labeled_frames[idx].instances = list()
>         labels.labeled_frames[idx].is_negative = True
> 
> sio.save_slp(labels, "labels_neg.v001.pkg.slp")
> > [@talmo](https://github.com/talmo):
> > We had a UX consideration that we wanted to address before doing this: how do you "mark" a frame as negative
> 
> [@talmo](https://github.com/talmo) [@roomrys](https://github.com/roomrys) [@gitttt-1234](https://github.com/gitttt-1234) Regarding the UX, it seems both [@skyshi0](https://github.com/skyshi0) and I spontaneously tried the same thing to label frames as negative: add an instance, and set all its points as hidden/invisible, so this might be the way to go. I'll try to analyse.
> 
> I can identify **three concerns**:
> 
> 1. Regardless of the UI, how clean-cut is the criterion itself?
> 2. Does this UI method of labelling frames as negative increase the likelihood of user error?
> 3. Is this UI method of labelling negative frames intuitive?
> 
> ##### Concern 1.
> If the assumption stands that a negative frame should have zero visible bodyparts and all frames that have zero visible bodyparts are negative frames, the following criterion seems straightforward, accurate, and easy to implement for automatic flagging of negative frames: if there is one or more instances and zero visible points across all of them, then and only then a frame is labelled as negative. This is also an easy test to put in label QC and `sio fix` (negative frame with one or more visible points being flagged as a point of concern).
> 
> ##### Concern 2.
> **_Assumption:_** Adding an instance and then hiding all of its points is a deliberate user action. If so, it follows that: **_Conclusion:_** Mislabelling a frame as negative in this manner is just as likely as any other labelling errors (e.g. switching symmetrical bodyparts, like left/right ears) - **no increase in the likelihood of erroneous labels**
> 
> ##### Concern 3.
> This is difficult to assess, but the first signal may be that two people independently tried to use the GUI in this way. However, if so, this would be pretty good news for Concern 2 as well. Tying this into Concern 2:
> 
> **_Assumption 1:_** If this method won't increase the likelihood of erroneous labels, and **_Assumption 2:_** If this method is intuitive, then it follows that: **_Conclusion:_** This way of labelling negative frames is **sufficient** in the GUI, with SLEAP-IO as a remaining, advanced option for more control.
> 
> Though I have no experience in UX, these questions are always super interesting to think about and I'm very curious to hear what you think.

Do you find flagging the frames to negative help reducing the chance to get the false positive predictions? And how does the script differ from adding the frames to the labeled frames and right clicking the nodes to "invisible"?

**Comment by davorvr** [comment]
> Do you find flagging the frames to negative help reducing the chance to get the false positive predictions? 

I just trained a bunch of models with different configs and backbones with a dataset that labels with `is_negative`. Since I generate that from a dataset with instances present, but no visible nodes, I'll try to run a few models with that one and let you know.

> And how does the script differ from adding the frames to the labeled frames and right clicking the nodes to "invisible"?

The script uses that information to figure out which frames are negative/background. The script will take any frame which contains one instance with all nodes hidden and will mark them as "negative". Currently, this special label can only be set using the `sleap_io` Python module and cannot be set via the GUI, but it's been introduced to the `.slp` format in SLEAP-IO version 0.6.3 (see [v0.6.3 release notes](https://github.com/talmolab/sleap-io/releases/v0.6.3)). If we explicitly mark frames as negative, they get specially categorised in `.slp` files, as [documented in SLEAP-IO docs](https://io.sleap.ai/latest/formats/slp/#negative-frames).

However, I tried searching `sleap-nn` code (the code responsible for actually training the networks), but I haven't been able to find details on how NN training treats frames labelled as negative, or even whether it recognises them as such since I haven't been able to find any mention of negative frames in `sleap-nn` code.

**Comment by talmo** [comment]
Hey guys,

Just wanted to circle back here -- we usually add support for data structures in sleap-io before we fully support training models with them in sleap-nn.

We're working on a PR to add support for this feature --> https://github.com/talmolab/sleap-nn/pull/484

Will update back here when we're done for folks who want to do early testing!

Cheers,

Talmo

**Comment by davorvr** [comment]
@talmo That's great news, thanks for the update!
